scenario/WithdrawScenario.ts

```diff
@@ -1,8 +1,80 @@
-import { scenario } from './context/CometContext';
+import { CometContext, scenario } from './context/CometContext';
 import { expect } from 'chai';
-import { expectApproximately, expectRevertMatches, getExpectedBaseBalance } from './utils';
+import { expectApproximately, getExpectedBaseBalance, isTriviallySourceable, isValidAssetIndex, NUM_ASSETS } from './utils';
+import { ContractReceipt } from 'ethers';
+
+async function testWithdrawCollateral(context: CometContext, assetNum: number): Promise<null | ContractReceipt> {
+  const comet = await context.getComet();
+  const { albert } = context.actors;
+  const { asset: assetAddress, scale: scaleBN } = await comet.getAssetInfo(assetNum);
+  const collateralAsset = context.getAssetByAddress(assetAddress);
+  const scale = scaleBN.toBigInt();
+
+  expect(await collateralAsset.balanceOf(albert.address)).to.be.equal(0n);
+  expect(await comet.collateralBalanceOf(albert.address, collateralAsset.address)).to.be.equal(100n * scale);
+
+  // Albert withdraws 100 units of collateral from Comet
+  const txn = await albert.withdrawAsset({ asset: collateralAsset.address, amount: 100n * scale })
+
+  expect(await collateralAsset.balanceOf(albert.address)).to.be.equal(100n * scale);
+  expect(await comet.collateralBalanceOf(albert.address, collateralAsset.address)).to.be.equal(0n);
+
+  return txn; // return txn to measure gas
+}
+
+async function testWithdrawFromCollateral(context: CometContext, assetNum: number): Promise<null | ContractReceipt> {
+  const comet = await context.getComet();
+  const { albert, betty } = context.actors;
+  const { asset: assetAddress, scale: scaleBN } = await comet.getAssetInfo(assetNum);
+  const collateralAsset = context.getAssetByAddress(assetAddress);
+  const scale = scaleBN.toBigInt();
+
+  expect(await collateralAsset.balanceOf(betty.address)).to.be.equal(0n);
+  expect(await comet.collateralBalanceOf(albert.address, collateralAsset.address)).to.be.equal(1000n * scale);
+
+  await albert.allow(betty, true);
+
+  // Betty withdraws 1000 units of collateral from Albert
+  const txn = await betty.withdrawAssetFrom({ src: albert.address, dst: betty.address, asset: collateralAsset.address, amount: 1000n * scale })
+
+  expect(await collateralAsset.balanceOf(betty.address)).to.be.equal(1000n * scale);
+  expect(await comet.collateralBalanceOf(albert.address, collateralAsset.address)).to.be.equal(0n);
+
+  return txn; // return txn to measure gas
+}
+
+for (let i = 0; i < NUM_ASSETS; i++) {
+  const amountToWithdraw = 100; // in units of asset, not wei
+  scenario(
+    `Comet#withdraw > collateral asset ${i}`,
+    {
+      filter: async (ctx) => await isValidAssetIndex(ctx, i) && await isTriviallySourceable(ctx, i, amountToWithdraw),
```

Comment:
So, why wouldn't we have a valid asset index?

Reply:
Oh, should `NUM_ASSETS` be `MAX_ASSETS`?

Reply:
Oh good point, yeah I'll rename it `MAX_ASSETS`.

Reply:
Done